### PR TITLE
Simplify dofmap collapsing and remove inconsistencies

### PR DESCRIPTION
**Pull request by garth-wells**
Avoids constructing new dofmap with new ordering when possible.

### CODE REVIEW

File: cpp/test/unit/common/IndexMap.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  //   else
+  //     throw std::runtime_error("Received unexpected data (2).");
+  // }
+
+} // namespace
+} // namespace
+
+TEST_CASE("Scatter using IndexMap", "[index_map_scatter]")
+{
+  CHECK_NOTHROW(test_scatter());
```
Comment (chrisrichardson) [inline cpp/test/unit/common/IndexMap.cpp]: This file seems to be commented out?

Comment (garth-wells) [reply cpp/test/unit/common/IndexMap.cpp]: I changed the interface, and it was too difficult to figure out what the test was doing.

Comment (chrisrichardson) [reply cpp/test/unit/common/IndexMap.cpp]: So delete it?

Comment (garth-wells) [reply cpp/test/unit/common/IndexMap.cpp]: Or you could add it back :)

Comment (garth-wells) [reply cpp/test/unit/common/IndexMap.cpp]: Added back.

**Pull request merged**